## 🔹 LÝ DO CHỌN THUẬT TOÁN KNN, CHIA DỮ LIỆU VÀ ĐÁNH GIÁ KẾT QUẢ

### 🧠 1. Lý do chọn thuật toán K-Nearest Neighbors (KNN)

Thuật toán **K-Nearest Neighbors (KNN)** được lựa chọn vì các lý do sau:

1. **Đơn giản và trực quan:**  
   KNN dựa trên nguyên lý “các điểm dữ liệu gần nhau thường có cùng nhãn”. Điều này rất phù hợp cho bài toán phân loại nhị phân như chẩn đoán bệnh tiểu đường (0: Không bị, 1: Bị).

2. **Không giả định phân phối dữ liệu:**  
   KNN **không yêu cầu dữ liệu tuân theo phân phối xác suất nào** (như Gaussian NB) nên thích hợp với dữ liệu y tế có tính biến động cao.

3. **Thích hợp khi dữ liệu được chuẩn hóa (scaled):**  
   Sau khi dữ liệu được **chuẩn hóa về cùng thang đo**, KNN hoạt động hiệu quả vì khoảng cách giữa các điểm được tính toán công bằng hơn.

4. **Dễ mở rộng và điều chỉnh:**  
   Tham số **K (số lượng láng giềng gần nhất)** cho phép điều chỉnh độ nhạy của mô hình. K nhỏ → nhạy hơn, K lớn → ổn định hơn.

---

### ⚙️ 2. Chia tập huấn luyện và kiểm thử

- Dữ liệu sau khi chuẩn hóa (`diabetes_dataset3_scaled.csv`) được chia theo tỷ lệ:
  - **Train-set:** 80% (614 mẫu) dùng để huấn luyện mô hình.  
  - **Test-set:** 20% (154 mẫu) dùng để đánh giá mô hình.  
- Cách chia này giúp mô hình có đủ dữ liệu học, đồng thời vẫn đảm bảo khả năng kiểm tra tính tổng quát (generalization).
- Dùng `random_state=42` để tái lập kết quả.

```python
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42)
---

In [1]:
# ===========================
# 03c_KNN_Training.ipynb
# ===========================

import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report, precision_score, recall_score, f1_score

# ===========================
# ĐỌC DỮ LIỆU
# ===========================
df3 = pd.read_csv("diabetes_dataset3_scaled.csv")

def evaluate_model(y_true, y_pred, model_name, dataset_name):
    acc = accuracy_score(y_true, y_pred)
    prec = precision_score(y_true, y_pred)
    rec = recall_score(y_true, y_pred)
    f1 = f1_score(y_true, y_pred)
    print(f"\n{model_name} ({dataset_name})")
    print(f"- Accuracy: {acc:.4f}")
    print(f"- Precision: {prec:.4f}")
    print(f"- Recall: {rec:.4f}")
    print(f"- F1: {f1:.4f}")
    print("\nConfusion Matrix:")
    print(confusion_matrix(y_true, y_pred))
    # return acc, prec, rec, f1

X = df3.drop('Outcome', axis=1)
y = df3['Outcome']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

k_values = [7, 9, 11, 13, 15, 17, 19, 21]
best_k, best_acc = 0, 0

for k in k_values:
    knn = KNeighborsClassifier(n_neighbors=k)
    knn.fit(X_train, y_train)
    acc = accuracy_score(y_test, knn.predict(X_test))
    print(f"K={k}: Accuracy={acc:.4f}")
    if acc > best_acc:
        best_acc, best_k = acc, k

print(f"\nBest K: {best_k} (Accuracy={best_acc:.4f})")
final_knn = KNeighborsClassifier(n_neighbors=best_k)
final_knn.fit(X_train, y_train)
y_pred = final_knn.predict(X_test)
evaluate_model(y_test, y_pred, f"KNN (K={best_k})", "Dataset 3 (Scaled)")


K=7: Accuracy=0.7143
K=9: Accuracy=0.7273
K=11: Accuracy=0.7597
K=13: Accuracy=0.7597
K=15: Accuracy=0.7403
K=17: Accuracy=0.7597
K=19: Accuracy=0.7727
K=21: Accuracy=0.7662

Best K: 19 (Accuracy=0.7727)

KNN (K=19) (Dataset 3 (Scaled))
- Accuracy: 0.7727
- Precision: 0.7174
- Recall: 0.6000
- F1: 0.6535

Confusion Matrix:
[[86 13]
 [22 33]]


### 📊 3. Kết quả thử nghiệm với các giá trị K khác nhau

| K | Accuracy |
|---|-----------|
| 7  | 0.7143 |
| 9  | 0.7273 |
| 11 | 0.7597 |
| 13 | 0.7597 |
| 15 | 0.7403 |
| 17 | 0.7597 |
| 19 | 0.7727 |
| 21 | 0.7662 |

➡️ **Giá trị K tối ưu:** `K = 19`  
➡️ **Độ chính xác cao nhất:** `Accuracy = 0.7727` (≈ 77.27%)

---

### 📈 4. Đánh giá chi tiết mô hình KNN (K = 19)

- **Accuracy:** 0.7727  
- **Precision:** 0.7174  
- **Recall:** 0.6000  
- **F1-score:** 0.6535  

**➡ Ma trận nhầm lẫn (Confusion Matrix):**

|                    | Dự đoán Không bị | Dự đoán Bị |
|--------------------|------------------|-------------|
| **Thực tế Không bị** | 86 | 13 |
| **Thực tế Bị**        | 22 | 33 |

---

### 🧩 5. Phân tích và nhận xét

1. **Hiệu suất tổng thể:**
   - Mô hình đạt **độ chính xác cao 77.27%**, cao hơn cả hai mô hình trước đó (ID3 và GaussianNB).  
   - **Precision = 0.7174** và **Recall = 0.6000** cho thấy KNN có khả năng dự đoán **tốt hơn về độ chính xác**, nhưng **vẫn bỏ sót một phần bệnh nhân thật sự bị tiểu đường**.

2. **Cân bằng giữa các chỉ số:**
   - **F1-score = 0.6535** là mức trung bình tốt, thể hiện sự cân bằng tương đối giữa Precision và Recall.  
   - Điều này nghĩa là KNN dự đoán “Bị” khá chính xác, nhưng vẫn cần cải thiện để nhận diện thêm các ca bệnh thật.

3. **Ảnh hưởng của giá trị K:**
   - Khi **K nhỏ** (như K=7), mô hình dễ bị nhiễu và không ổn định.  
   - Khi **K quá lớn**, mô hình có xu hướng trung bình hóa kết quả.  
   - Giá trị **K=19** là điểm cân bằng hợp lý, giúp mô hình vừa ổn định vừa chính xác hơn.

4. **Ý nghĩa từ Confusion Matrix:**
   - Mô hình **phân loại đúng 86/99 trường hợp không bị** (tỷ lệ 86.9%).  
   - Với các ca “Bị”, mô hình **phát hiện đúng 33/55 trường hợp** (60%), cho thấy vẫn còn **40% bệnh nhân bị tiểu đường bị dự đoán sai là không bị**.  
   - Trong y tế, đây là vấn đề cần cân nhắc — mô hình có thể **ưu tiên Recall** hơn trong các ứng dụng thực tế để giảm thiểu bỏ sót ca bệnh.

5. **Ưu điểm của KNN:**
   - Dễ cài đặt, không cần huấn luyện phức tạp.  
   - Là mô hình **phi tham số (non-parametric)**, thích hợp khi không biết rõ phân phối dữ liệu.  
   - Khi dữ liệu đã được chuẩn hóa, KNN thường cho kết quả **ổn định và trực quan**.

6. **Hạn chế:**
   - **Chi phí tính toán cao** khi dữ liệu lớn, do phải đo khoảng cách đến tất cả điểm huấn luyện.  
   - **Nhạy cảm với nhiễu** và đặc trưng không liên quan.  
   - Không giải thích rõ ràng mối quan hệ giữa các biến đầu vào và đầu ra (mô hình hộp đen).

---

### 🧠 6. Kết luận

- Mô hình **KNN (K=19)** đạt **độ chính xác cao nhất (77.27%)** trong ba mô hình đã thử nghiệm.  
- Mô hình thể hiện khả năng **phân loại khá tốt**, đặc biệt là nhận diện chính xác phần lớn các bệnh nhân “không bị tiểu đường”.  
- Tuy nhiên, để ứng dụng trong thực tế y tế, cần cải thiện khả năng phát hiện bệnh nhân thực sự bị (tăng Recall).  

➡ Có thể cải thiện bằng cách:
- Thử **gán trọng số theo khoảng cách** (`weights='distance'`).
- Dùng **GridSearchCV** để tìm giá trị K tối ưu chính xác hơn.  
- Kết hợp với **feature selection hoặc PCA** để giảm nhiễu.  

---

> ✅ **Tóm lại:**  
> Mô hình **KNN với K=19** mang lại hiệu suất tốt nhất trong các thử nghiệm, với độ chính xác ~77% và khả năng phân loại cân bằng.  
> Đây là lựa chọn hiệu quả cho bài toán dự đoán tiểu đường trên dữ liệu đã chuẩn hóa, đồng thời là nền tảng tốt để phát triển và tối ưu hóa mô hình trong các bước tiếp theo.
